sql-plugin/src/main/scala/com/nvidia/spark/rapids/conditionalExpressions.scala

```diff
@@ -82,8 +108,151 @@ case class GpuIf(
+   *
+   * This method performs lazy evaluation on the GPU by first filtering the
+   * input batch into two batches - one for rows where the predicate is true
+   * and one for rows where the predicate is false. The expressions are
+   * evaluated against these batches and then the results are combined
+   * back into a single batch using the gather algorithm.
+   */
+  private def conditionalWithSideEffects(
+      batch: ColumnarBatch,
+      pred: GpuColumnVector,
+      gpuTrueExpr: GpuExpression,
+      gpuFalseExpr: GpuExpression): GpuColumnVector = {
+
+    val colTypes = GpuColumnVector.extractTypes(batch)
+
+    withResource(GpuColumnVector.from(batch)) { tbl =>
+      withResource(pred.getBase.unaryOp(UnaryOp.NOT)) { inverted =>
+        // evaluate true expression against true batch
+        val tt = withResource(filterBatch(tbl, pred.getBase, colTypes)) { trueBatch =>
+          gpuTrueExpr.columnarEval(trueBatch)
+        }
+        withResourceIfAllowed(tt) { _ =>
+          // evaluate false expression against false batch
+          val ff = withResource(filterBatch(tbl, inverted, colTypes)) { falseBatch =>
+            gpuFalseExpr.columnarEval(falseBatch)
+          }
+          withResourceIfAllowed(ff) { _ =>
+            val finalRet = (tt, ff) match {
+              case (t: GpuColumnVector, f: GpuColumnVector) =>
+                withResource(gather(pred.getBase, t)) { trueValues =>
+                  withResource(gather(inverted, f)) { falseValues =>
+                    pred.getBase.ifElse(trueValues.getColumn(0), falseValues.getColumn(0))
+                  }
+                }
+              case (t: GpuScalar, f: GpuColumnVector) =>
+                withResource(gather(inverted, f)) { falseValues =>
+                  pred.getBase.ifElse(t.getBase, falseValues.getColumn(0))
+                }
+              case (t: GpuColumnVector, f: GpuScalar) =>
+                withResource(gather(pred.getBase, t)) { trueValues =>
+                  pred.getBase.ifElse(trueValues.getColumn(0), f.getBase)
+                }
+              case (_: GpuScalar, _: GpuScalar) =>
+                throw new IllegalStateException(
+                  "scalar expressions can never have side effects")
+            }
+            GpuColumnVector.from(finalRet, dataType)
+          }
+        }
+      }
+    }
+  }
+
+  private def filterBatch(
+      tbl: Table,
+      pred: ColumnVector,
+      colTypes: Array[DataType]): ColumnarBatch = {
+    withResource(tbl.filter(pred)) { filteredData =>
+      GpuColumnVector.from(filteredData, colTypes)
+    }
+  }
+
+  private def boolToInt(cv: ColumnVector): ColumnVector = {
+    withResource(GpuScalar.from(1, DataTypes.IntegerType)) { one =>
+      withResource(GpuScalar.from(0, DataTypes.IntegerType)) { zero =>
+        cv.ifElse(one, zero)
+      }
+    }
+  }
+
+  private def gather(predicate: ColumnVector, t: GpuColumnVector): Table = {
```

Comment:
Nit: I think the code gets a bit cleaner if this returned a `ColumnVector` rather than a `Table`.  It's a bit weird to see a gather on a column return a table, and every caller ends up having to call `.getColumn(0)` on the resulting table since everyone wants to treat the result as just a column.  This method can `incRefCount` the column to allow closing the outer `Table` yet keep the underlying column alive to be returned.